application/org.openjdk.jmc.flightrecorder.flameview/src/main/java/org/openjdk/jmc/flightrecorder/flameview/views/FlameGraphView.java

```diff
@@ -331,44 +338,66 @@ public void saveState(IMemento memento) {
 	public void selectionChanged(IWorkbenchPart part, ISelection selection) {
 		if (selection instanceof IStructuredSelection) {
 			Object first = ((IStructuredSelection) selection).getFirstElement();
-			setItems(AdapterUtil.getAdapter(first, IItemCollection.class));
+			IItemCollection items = AdapterUtil.getAdapter(first, IItemCollection.class);
+			if (items == null) {
+				rebuildModel(ItemCollectionToolkit.build(Stream.empty()));
+			} else if (!items.equals(currentItems)) {
+				rebuildModel(items);
+			}
 		}
 	}
 
-	private void setItems(IItemCollection items) {
-		if (items != null) {
-			currentItems = items;
-			rebuildModel(items);
+	private void rebuildModel(IItemCollection items) {
+		if (modelCalculationFuture != null) {
+			modelCalculationFuture.cancel(true);
 		}
+
+		modelCalculationFuture = getModelPreparer(items, true);
+		currentItems = items;
 	}
 
-	private void rebuildModel(IItemCollection items) {
-		// Release old model before building the new
-		if (currentModelCalculator != null) {
-			currentModelCalculator.cancel(true);
-		}
-		currentModelCalculator = getModelPreparer(items, frameSeparator, true);
-		currentModelCalculator.thenAcceptAsync(this::setModel, DisplayToolkit.inDisplayThread())
-				.exceptionally(FlameGraphView::handleModelBuildException);
+	private StacktraceModel createStacktraceModel(IItemCollection items) {
+		return new StacktraceModel(threadRootAtTop, frameSeparator, items);
 	}
 
-	private CompletableFuture<TraceNode> getModelPreparer(
-		final IItemCollection items, final FrameSeparator separator, final boolean materializeSelectedBranches) {
-		return CompletableFuture.supplyAsync(() -> {
-			StacktraceModel model = new StacktraceModel(threadRootAtTop, frameSeparator, items);
-			TraceNode root = TraceTreeUtils.createRootWithDescription(items, model.getRootFork().getBranchCount());
-			return TraceTreeUtils.createTree(root, model);
-		}, MODEL_EXECUTOR);
+	private Future<Void> getModelPreparer(final IItemCollection items, final boolean materializeSelectedBranches) {
+
+		final Callable<Void> modelPreparerTask = () -> {
+			// reduce the pressure on displayThread during a scrolling, ensure movement is final, seems there could be visibility issue
+			Thread.sleep(200);
+			StacktraceModel model = createStacktraceModel(items);
+			Fork rootFork = model.getRootFork();
+			if (materializeSelectedBranches) {
+				Branch selectedBranch = StacktraceModel.getLastSelectedBranch(rootFork);
+				if (selectedBranch != null) {
+					selectedBranch.getEndFork();
+				}
+			}
+
+			TraceNode root = TraceTreeUtils.createRootWithDescription(items, rootFork.getBranchCount());
+			if (Thread.currentThread().isAlive()) {
```

Comment:
I don’t get these thread changes. The current thread is executing your code, and is therefore always alive. Also, if you want to make these calculations abortable, I’d much prefer an API for it. Maybe an Abortable interface, or perhaps even better, providing some sort of a progress interface where progress could be reported to and checked against whether it should abort every now and then. We can discuss which alternative makes most sense.

Reply:
I've reconsidered my thoughts. I wanted to cancel the Task calculations inside the executor thread before it get's jointed with the main. I've added the Scheduler executor and a bit delay its start, in case of fast scrolling, it's more transparent. I've also cleared the code.